compiler/rustc_metadata/src/native_libs.rs

```diff
@@ -253,15 +263,46 @@ impl Collector<'tcx> {
             )
             .emit();
         }
-        if lib.kind == NativeLibKind::RawDylib && !self.tcx.features().raw_dylib {
-            feature_err(
-                &self.tcx.sess.parse_sess,
-                sym::raw_dylib,
-                span.unwrap_or(rustc_span::DUMMY_SP),
-                "kind=\"raw-dylib\" is unstable",
-            )
-            .emit();
+        // this just unwraps lib.name; we already established that it isn't empty above.
+        if let (NativeLibKind::RawDylib, Some(lib_name)) = (lib.kind, lib.name) {
+            if !self.tcx.features().raw_dylib {
+                feature_err(
+                    &self.tcx.sess.parse_sess,
+                    sym::raw_dylib,
+                    span.unwrap_or(rustc_span::DUMMY_SP),
+                    "kind=\"raw-dylib\" is unstable",
+                )
+                .emit();
+            } else {
```

Comment:
Feature gates should only be used to report errors and not change behaviour. It's fine to keep just move everything out of the else branch there and just continue to report any other errors.

Reply:
It sounds like we've got conflicting requirements here, unfortunately.

Earlier feedback asked me to make the feature-gating test run on all platforms, which I have done.  Moving the other error checks (back) out of the `else` and continuing to report other errors means that the feature-gating test would have different expected behavior on different platforms, as the error checks in the else branch are (necessarily) platform-specific.  I was not able to find a way to construct a test whose expected output depends on the platform, so it would seem that we have two options here: either leave things as they are, or split the feature-gating test into several platform-specific variants.

Are there other possibilities that I might have missed?

Reply:
You can use [revisions](https://rustc-dev-guide.rust-lang.org/tests/adding.html#revisions) to use a single test file for multiple tests. The [inline asm tests](https://github.com/rust-lang/rust/blob/af2ed1b51887ede4d766ee123d3fa216970d1fdd/src/test/ui/asm/inline-syntax.rs) are a good example. Then you'll have different .stderr files like: `my-test.<rev>.stderr`.

Admittedly that functionality is a lil limited so I'm fine if it's not possible. I'll defer to someone from the compiler team on I guess what the official policy is. I'm just going off comments I received on my own PRs before.

Reply:
Ah, I hadn't seen the revision mechanism before, thanks.

I'm not sure I see a good way of making that work here, though.  Based on the inline asm tests, it looks like I'd have to do something like
```
// revisions: windows-msvc windows-gnu other
// [windows-msvc] compile-flags: --target x86_64-pc-windows-msvc
// [windows-gnu] compile-flags: --target x86_64-pc-windows-gnu
// [other] compile-flags: --target x86_64-unknown-linux-gnu
```
The problem is that, if I understand correctly, I have to specify a target string for each revision.  Unless we want to add revisions for all possible targets (which would add a lot of clutter and non-trivial maintenance overhead), we're back to the problem of the feature-gating test only running on specific architectures.

Reply:
>make the feature-gating test run on all platforms

Is this some tidy restriction?
I would personally put the "only supported on Windows" before the feature gate error, because it doesn't make sense to report the feature gate error for something that's going to error later anyway with 100% guarantee.

Reply:
OK, I've reordered the error checks.

This requires changing the feature-gating tests back to windows-only (one for msvc, one for gnu), which I've also done in the most recent push.  This doesn't bother me; just wanted to call it out, since it's apparently a departure from the norm.